jaxopt/_src/levenberg_marquardt.py

```diff
@@ -0,0 +1,356 @@
+                                    increase_factor, residual, gradient,
+                                    updated_params)
+
+      # Calling the jax condition function:
+      # Note that only the parameters that are used in the rest of the program
+      # are returned by this function.
+
+      params, error, grad_error, damping_factor, increase_factor, gradient, residual = (
+          self.update_state_using_gain_ratio(gain_ratio,
+                                             gain_ratio_test_init_state, *args,
+                                             **kwargs))
+
+      return params, error, grad_error, damping_factor, increase_factor, gradient, residual
+
+    params, error, grad_error, damping_factor, increase_factor, gradient, residual = (
+        jax.lax.cond(
+            small_dparams_test_pred(dparams_diff), small_dparams_test_true_func,
+            small_dparams_test_false_func, *small_dparams_test_init_state))
+
+    return params, error, grad_error, damping_factor, increase_factor, gradient, residual
+
+  def update(self, params, state: NamedTuple, *args, **kwargs) -> base.OptStep:
+    """Performs one iteration of the least-squares solver.
+
+    Args:
+      params: pytree containing the parameters.
+      state: named tuple containing the solver state.
+
+    Returns:
+      (params, state)
+    """
+
+    # Current value of the loss function F=0.5*||f||^2.
+    loss_curr = state.loss
+
+    def jtj_reg_matvec(vec):
+      """The (J^T.J + damping_factor*I) matrix vector product operator."""
+      return self._jtj_reg_op(params, state.damping_factor, vec, *args,
+                              **kwargs)
+
+    matvec = lambda v: self._jtj_op(params, v, *args, **kwargs)
+
+    if isinstance(self.solver, Callable):
+      delta_params = self.solver(
+          jtj_reg_matvec, state.gradient, init=state.delta)
+    elif self.solver == 'cg':
+      delta_params = solve_cg(
+          matvec, state.gradient, ridge=state.damping_factor, init=state.delta)
+    elif self.solver == 'bicgstab':
+      delta_params = solve_bicgstab(
+          matvec, state.gradient, ridge=state.damping_factor, init=state.delta)
+    elif self.solver == 'gmres':
+      delta_params = solve_gmres(
+          matvec, state.gradient, ridge=state.damping_factor, init=state.delta)
+    elif self.solver == 'cholesky':
+      delta_params = solve_cholesky(
+          matvec, state.gradient, ridge=state.damping_factor)
+    elif self.solver == 'inv':
+      delta_params = solve_inv(
+          matvec, state.gradient, ridge=state.damping_factor)
+    delta_params = -delta_params
+
+    # Define the functions required for the major conditional of the algorithm,
+    # which checks the magnitude of dparams and checks if it is small enough.
+    # for the value of dparams.
+
+    tree_mul_term = tree_mul(self.xtol,
+                             tree_add(tree_l2_norm(params), self.xtol))
```

Comment:
Same here and below. Normally `tree_l2_norm` returns a scalar so you don't need to use tree utils on top.

Reply:
Done. thanks for catching.